### ISSUE TITLE
Drop-In UI hangs on returning, after attempting to add a duplicate payment method via the Card Details screen.

### ISSUE THREAD

**Issue by soarb** [braintree/braintree-android-drop-in#357]
### General information

* SDK/Library version: com.braintreepayments.api:drop-in:6.2.0
* Environment: Sandbox
* Android Version and Device: Samsung Galaxy S10 with Android 12.0

### Issue description

The experience between the iOS drop-in and Android drop-in is quite large. This issue is focussing on the drop-in experience on Android when adding an already existing test payment method.

On Android, tapping "Credit or Debit Card" from the drop-in menu presents the Card Details screen. Unlike the iOS version, it doesn't automatically display a keyboard to enter the Card Number even though this field has focus.

Having added the (already existing) test card 4000 0000 0000 1000, tapping next allows me to enter the Expiration Date, which according to [this](https://developer.paypal.com/braintree/docs/guides/3d-secure/testing-go-live) list should be 01/25 or 01/2025 (I admit in the video links I posted I'm entering 01/23!)

Tapping ADD CARD shows a spinner for a brief moment. No feedback from the UI. Can repeatedly tap ADD CARD with no response from the UI other than the spinner.

The only way to progress from this screen is to tap back, however tapping back now actually reveals the "Card number is invalid" error. I'd like to point out that on iOS, the error message is actually more useful i.e., "This credit card already exists as a saved payment method" which is FAR more useful.

Now if I decide that I don't want to continue with entering the details for this payment method, tapping back returns me to the initial Drop-In screen with a permanent spinner which doesn't resolve to anything.

Please see the full [android experience](https://youtu.be/scCdVRhx_os)

In comparison, this is the [iOS experience](https://youtube.com/shorts/WXwJcB8HuSI) which feels far slicker in comparison.

Would appreciate some info regarding why Android behaves the way it does compared to the iOS version (maybe it's due to the Sandbox vs Production environment?) and more specifically how to resolve the infinite spinner when returning to the main Drop-In UI after failing to add a new card.

We're looking to release a new version of our App quite soon to get away from the deprecated version we're migrating from, but with small Android experiences like this, I'm not sure we can.

Any advice would be gratefully appreciated as always :)

Cheers

Benj

**Comment by sshropshire** [comment]
Hey @soarb thanks for this feedback. I'll look into this. Out of curiosity, are you able to open a network profiler and inspect the HTTP responses that come from the gateway when tokenizing a card? Also is this 100% reproducible?

**Comment by soarb** [comment]
hi @sshropshire ... it's definitely 100% reproducible, first thing I'll do tomorrow is recreate it whilst checking the HTTP responses and get back to you :)

**Comment by soarb** [comment]
Hi @sshropshire. At the point I attempt to add the duplicate (test) payment method, I can see a GraphQL network request go out to https://payments.sandbox.braintree-api.com/graphql with the following payload:-

```
{
  "clientSdkMetadata": {
    "platform": "android",
    "sessionId": "07b17a1889a94798add54b84a2771402",
    "source": "form",
    "integration": "custom"
  },
  "query": "mutation TokenizeCreditCard($input: TokenizeCreditCardInput!) {  tokenizeCreditCard(input: $input) {    token    creditCard {      bin      brand      expirationMonth      expirationYear      cardholderName      last4      binData {        prepaid        healthcare        debit        durbinRegulated        commercial        payroll        issuingBank        countryOfIssuance        productId      }    }  }}",
  "operationName": "TokenizeCreditCard",
  "variables": {
    "input": {
      "options": {
        "validate": true
      },
      "creditCard": {
        "number": "4000000000001000",
        "expirationMonth": "01",
        "expirationYear": "25"
      }
    }
  }
}
```

The response is as expected (same as on iOS) i.e.,

```
{
  "errors": [
    {
      "message": "Duplicate card exists in the vault",
      "locations": [
        {
          "line": 1,
          "column": 66
        }
      ],
      "path": [
        "tokenizeCreditCard"
      ],
      "extensions": {
        "errorClass": "VALIDATION",
        "errorType": "user_error",
        "inputPath": [
          "input",
          "creditCard",
          "number"
        ],
        "legacyCode": "81724"
      }
    }
  ],
  "data": {
    "tokenizeCreditCard": null
  },
  "extensions": {
    "requestId": "61937b8b-7513-49ce-831b-ab03e83ef604"
  }
}
```

but the Drop-In UI's Card Details screen does absolutely nothing with it. I'd at least expect it to show an Alert stating that a "Duplicate card exists in the vault"?

Tapping the back arrow does nothing other than display the same Card Details screen but _now_ the "Card number is invalid" error shows below the "Card Number" field.

Again, no error to confirm that a duplicate payment method exists (as the GraphQL response correctly indicates).

Tapping back _again_ leaves me on the initial Drop-In UI screen i.e., it _should_ display my "Recent" payment methods and allow me to switch to any of the "Other" payment methods, but all I'm left with is a blank Drop-In UI with a spinner.

Absolutely no network requests get sent out at this point.

The only way to recover from this state is to dismiss the Activity with the hardware back button.

Cheers

Ben

**Comment by soarb** [comment]
@sshropshire further to this, and just an observation (not strictly a bug), but when managing multiple payment methods via the drop-in and deleting all of them, returning to the main drop-in screen to "Select Payment Method" displays no previous payment methods as expected, but the height of the drop-in is as if that element is still present, i.e., the form hasn't collapsed back down to it's correct (content) size:-

![IMG_2309](https://user-images.githubusercontent.com/316338/184129171-e16254ac-d1b1-403d-ae3b-5659477b948a.jpg)

**Comment by sshropshire** [comment]
@soarb I can see the difference in response handling we may need to provide a fix for this in our core repo. The visual bug is also interesting we may have to re-measure the height of the bottom sheet in that case.

Thanks for your help and feedback! Also we have a PR up for #358 to remove `DropInRequest` from the `DropInClient` constructor. Once approved, we'll be able to add this in the next minor version release.

As we're working on some of these more critical issues, feel free to add additional feature requests etc. via GitHub issues if you have any. We're starting to prioritize more enhancements in the coming months for our next major version of the Android SDK.

**Comment by soarb** [comment]
@sshropshire great stuff! really appreciate all the hard work you and the team are putting into this, especially removing the DropInRequest from the DropInClient constructor :)

**Comment by sshropshire** [comment]
@soarb this is now released in version [6.4.0](https://github.com/braintree/braintree-android-drop-in/blob/master/CHANGELOG.md#640). If you have additional concerns, feel free to open another issue and we'll take a look. We are also tracking the visual payment sheet bug you mentioned and hope to have a fix soon.
